shadowsocks-libev/files/shadowsocks.init

```diff
@@ -1,298 +1,512 @@
+    # If the route is not in the routing table, we don't want to add a bypass rule for ss-redir
+    # An example of wrong route is 9.9.9.9/30. The correct one is 9.9.9.8/30.
+    if [ "$NETWORK" != "$IP" ]; then
+        _log 4 "Skipping static route with wrong network ($IP/$PREFIX instead of $NETWORK/$PREFIX)"
+        return 0
+    fi
+
+    # If the route made it here and survived, let's add it! :)
+    ipset -q add socks_staging "$NETWORK/$PREFIX"
+}
 
-	config_load mwan3
-	config_foreach genrateRules rule
+# Local nets are already bypassed. However some clients may have public IPs directly on a WAN interface
+# This may happen for example with a modem in bridge mode. We want to bypass shadowsocks for each WAN address.
+# This avoids infinite network loop
+_populate_ipset_socks_bypass_wan_one() {
+    local ifname subnet
 
-	config_load network
-	generateLocalRules
-	config_foreach generateWanRoutesRules interface
-	config_foreach generateStaticRoutesRules route
+    # Get the current interface name
+    config_get ifname $1 ifname
 
-	# Swap ipset
-	ipset swap shadowsocks_reload shadowsocks_localnetwork
-	ipset destroy shadowsocks_reload
+    # Fetch IP/Netmask but skip interfaces with no IP
+    network_get_subnet subnet $ifname || return 0
+
+    # Find the network address and check everything is valid :)
+    eval "$(ipcalc.sh $subnet)"
+    if [ "$NETWORK" = "0.0.0.0" ] || [ "$PREFIX" = "0" ]; then
+        _log 4 "Skipping WAN IP ($NETWORK/$PREFIX) that I don't understand..."
+        return 0
+    fi
+
+    # Add the WAN IP to the exceptions list
+    ipset -q add socks_staging "$NETWORK/$PREFIX"
 }
 
-stop_service() {
-	shadowsocks_disable_interfaces_hook
-	# Clean shadowsocks tables
-	$IPTN -F SOCKS
-	$IPTN -X SOCKS
-	for ipset in $(ipset -n list | grep shadowsocks_); do
-		ipset destroy $ipset
-	done
-	# Clean up monitoring_ip rules
-	iptables-save | grep "shadowsocks_tracker" | grep "REDIRECT --to-ports" | sed "s/-A //g" | while read LINE; do
-		$IPTN -D $LINE
-	done
+# This updates the socks_hook chain to route or stop routing packets to shadowsocks
+# This is handled in a by-interface basis.
+_update_or_enable_hook() {
+    # First, reset the chain
+    _disable_hook
+
+    # For each zone, find if we got a lan zone
+    # And activate the routing to shadowsocks for all its interfaces
+    config_load firewall
+    config_foreach _update_or_enable_hook_one_zone zone
 }
 
-ifup() {
-	shadowsocks_enable_interfaces_hook "$@"
+_disable_hook() {
+    $IPT -F socks_hook
 }
 
-ifdown() {
-	shadowsocks_disable_interfaces_hook "$@"
+# $1 is a ref to the current zone section being processed
+_update_or_enable_hook_one_zone() {
+    # Let's fetch the current zone name
+    local zone
+    config_get zone $1 name
+
+    # Only traffic coming from a lan zone's interface will be routed to shadowsocks
+    if [ "$zone" = "lan" ]; then
```

Comment:
i sugesst to add an uci shadowsock list entry to add more zone than "lan", where value the lan is is the default.